### PR TITLE
Project template should allow for switching to WinUI 3 at build time

### LINKED ISSUE #6993: Make usewinui3 a build-time switch for app/libs
Today when a new app or lib is created, you have to choose between system xaml (default) or winui3.
We should instead write the scaffolding in a way that allows esp. modules to dual-compile as a way to soften the blow of an API break w.r.t. xaml --> winui

### PR DESCRIPTION
**Pull request by asklar**
Fixes #6993 #7077 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7014)

### CODE REVIEW

**Comment by dannyvv** [comment]
Shouldn't you make the PackageReference conditional as well?

**Comment by asklar** [comment]
> Shouldn't you make the PackageReference conditional as well?

https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#adding-a-packagereference-condition
> However, at presently, only the TargetFramework variable is supported.

**Comment by dannyvv** [comment]
> > However, at presently, only the TargetFramework variable is supported.
Booo <insert Statler and Waldorf gif>

File: vnext/template/cpp-app/src/pch.h
Comment (NickGerleman) [inline vnext/template/cpp-app/src/pch.h:9]: This seems like it would balloon template PCH. Can we avoid this? Seems likely to cause customers worse issues for compiler OOM, potentially perf.

File: vnext/template/cpp-app/src/App.h
Comment (NickGerleman) [inline vnext/template/cpp-app/src/App.h:5]: Same comment as in PCH. This is chunky. Would be best to include individual headers if possible.

File: vnext/template/cpp-app/src/pch.h
Comment (NickGerleman) [reply vnext/template/cpp-app/src/pch.h:9]: I think this is smaller than it used to be though... Need to be careful here though.

**Comment by NickGerleman** [comment]
If we want to add more CLI time build logic manipulation, are there other abstractions we could use than to tie to auto-linking? E.g. would we want this to run with autolink commands, generate autolinking warnings, etc?

Seems like the two share commonalities, and share code, but I'm not sure it makes sense for autolinking to do this work as well.

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:683]: This will break for PNPM or repos with multiple RNW versions, since the CLI does not depend on react-native-windows.

Need to search starting from project root. There should be other examples of this in the CLI, and I think this file might even have a `resolveTemplateDirectory` function or something like that.

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:739]: This file isn't in great shape for this, but it would be great to avoid new 100 line functions.

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:595]: Needs unit tests

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:657]: Does this have the potential to conflict with template packages.config?

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:608]: Does not handle XML comments

**Review by NickGerleman — CHANGES_REQUESTED** [review]
- Needs tests
- Ideally we shouldn't be doing custom string parsing for XML. Some bugs with current impl
- Breaks with PNPM/mutli-rn repo

Greater architectural questions on whether we should start doing generic fixup inside of autolinking. This has the potential rn to add some confusing errors, have some suprisinng behavior with autolinking flags also effecting unrelated concerns.

Please check to make sure this doesn't affect template PCH. E.g. before and after number.

File: vnext/template/cpp-app/src/App.h
Comment (NickGerleman) [reply vnext/template/cpp-app/src/App.h:5]: This would also leak namespace redefinition into the user project.

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (asklar) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:739]: please file a bug to refactor this file

**Review by NickGerleman — APPROVED** [review]
Implementation looks better to me now.

If this wasn't on your consideration list yet, the interaction between templating at init + formatting init files could be tricky. The scenario where:
1. Template packages.config is different after formatted
2. Build new project with VS
3. VS runs auto-link check
4. Build fails because autolinked outputs mismatched from template

Would appreciate testing that scenario, or identifying if we have coverage.

Re architecture, this is getting a bit messier than I would like. It feels like we want a general purpose layer for build-time code-generation. I.e. it doesn't really make sense to have the autolink command trigger this, or show autolinking failure messages if this fails.

I don't think this PR is a good terminal state, but it seems like a bridge we can iterate on.

Will defer to @jonthysell on his thoughts

**Comment by asklar** [comment]
> > > However, at presently, only the TargetFramework variable is supported.
> 
> Booo [insert Statler and Waldorf gif]

@dannyvv good news I figured out a way to make it conditional without using `Condition`!

File: packages/@react-native-windows/cli/src/e2etest/projectConfig.test.ts
Comment (dannyvv) [inline packages/@react-native-windows/cli/src/e2etest/projectConfig.test.ts:141]: @NickGerleman/@asklar/@jonthyssel: Should we consider the same pattern for nuget for community modules?

Comment (dannyvv) [reply packages/@react-native-windows/cli/src/e2etest/projectConfig.test.ts:141]: And hopefully make autolink aware of this...
@asklar: Similar for community modules that support both winui2/winui3

Comment (asklar) [reply packages/@react-native-windows/cli/src/e2etest/projectConfig.test.ts:141]: @dannyvv could you clarify what you mean? this file is a test

File: vnext/template/shared-lib/proj/BuildFlags.props
Comment (dannyvv) [inline vnext/template/shared-lib/proj/BuildFlags.props:6]: Would useNuget be appropriate as well?

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (jonthysell) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:596]: I see that the slnFile is fully resolved, but we should be able to just pull it out of the config, especially if we're exporting this function to be used outside of autolinking.

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (jonthysell) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:634]: Can this method be rename specifically to something about selecting the right Xaml version?

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (jonthysell) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:604]: Do we have to support setting the flag in two places? Can we just take the value from react-native.config.js (which rather than ever being undefined, has a default if not specified) and always override what's in BuildFlags.props?

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (jonthysell) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:675]: Is there any way to extract out steps for adding/updating/removing nodes in a packages.config into their own utility functions? Then I could re-use it when autolinking needs to do the same.

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (jonthysell) [inline packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: Is there some existing "pretty xml" code we could call here rather than rolling our own?

**Review by jonthysell — CHANGES_REQUESTED** [review]

**Review by jonthysell — APPROVED** [review]
Whoops, didn't see the PR #7060 for refactoring autolinking. I guess you can probably address my comments there.

File: vnext/template/shared-lib/proj/BuildFlags.props
Comment (NickGerleman) [reply vnext/template/shared-lib/proj/BuildFlags.props:1]: This was in the app template before.

This one is a bit special. Meant to allow customizing Microsoft.ReactNative source build. The project looks in solutionDir for this file, to change its own behavior.

Added to allow specialized compilation for Hermes, since we cannot yet enable its conditional compilation in MSRN by default.

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (asklar) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: I tried using xml-formatter but it doesn't work with ES6 apparently. Filed https://github.com/chrisbottin/xml-formatter/issues/31 but I didn't want to block on it :)

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (asklar) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:604]: that would mean that if you want winui3 you _have_ to have a react-native.config.js which seemed more of a divergence than I felt comfortable with, but we can discuss more if you feel strongly about it :)

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: Please remove package from "package.json" if no longer being used.

That specific warning is actually more of an issue on our end not supporting certain libraries with ES6 default export.

Depending on usage, there is a syntactic fix, or a possible TypeScript compiler config change. I've been meaning to look into that one though.

Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: Yeah, I think that lib should work. Just make sure you do this:

`import * as formater from 'xml-formatter'`
instead of 
`import formatter from 'xml-formatter'`

Comment (asklar) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: yeah I had done that but then I get an error when trying to run tests about some babel thing not supporting something else. I'll file a follow up for it

File: vnext/template/shared-lib/proj/BuildFlags.props
Comment (asklar) [reply vnext/template/shared-lib/proj/BuildFlags.props:1]: yep, pre-existing art :) we can revisit if need be. It's nice to have experimental features be gated by a file that's just that.

File: vnext/template/cs-app/proj/MyApp.csproj
Comment (asklar) [reply vnext/template/cs-app/proj/MyApp.csproj:174]: there's an official ask on devdiv to support packagereferences in vcxproj but no commitment yet AFAICT

File: vnext/template/shared-lib/proj/BuildFlags.props
Comment (NickGerleman) [reply vnext/template/shared-lib/proj/BuildFlags.props:1]: https://github.com/microsoft/react-native-windows/pull/7075

File: vnext/template/cpp-lib/src/pch.h
Comment (asklar) [reply vnext/template/cpp-lib/src/pch.h:9]: This file lives in Microsoft.ReactNative.Cxx. Native modules that don't need to instantiate anything in xaml or DComp can delete

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: @asklar the babel error is if you do `import x = require('x')`. Could you please try with `import * as x from 'x'` instead? There will be a different error if there is one.

Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: Worst case, you can just `require` the thing instead of import, with the consequence of losing typings.

File: vnext/template/shared-lib/proj/BuildFlags.props
Comment (asklar) [reply vnext/template/shared-lib/proj/BuildFlags.props:6]: If usenuget needs build-time behavior, sure. I'm going to keep this PR focused on WinUI though, feel free to add to the template if you have a use for an additional flag :)

File: vnext/template/cs-app/src/App.xaml.cs
Comment (asklar) [reply vnext/template/cs-app/src/App.xaml.cs:2]: I'm assuming you mean flip the order, in which case sure :)

File: vnext/template/shared-app/proj/BuildFlags.props
Comment (asklar) [reply vnext/template/shared-app/proj/BuildFlags.props:1]: I can see a reason to use a props file, it makes build logic simpler (just use whatever it says in the props file), versus having the CLI parse react-native.config.js and then pass /p:foo=bar to msbuild.
Plus this allows VS builds to work too, whereas relying on react-native.config.js does not allow those to work (without some work to project values from the config to some props file)

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (asklar) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: image

File: vnext/template/cpp-app/proj/MyApp.vcxproj
Comment (asklar) [reply vnext/template/cpp-app/proj/MyApp.vcxproj:3]: the CSharpApp.props etc. need the value of UseWinUI3 to be already defined

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: Thanks.

It looks like I misread the typings for the library as `module.exports = format` instead of `export = format`. The difference between those is that the former defines a JS interface (transpiled) that refers to CommonJS semantics. The latter defines a JS interface (transpiled) with ES6 semantics.

Would you be able to try `const formatter = require('xml-formatter').__esModule`?

The import will be non-type-safe. You can still import typings from the module though.

We can fix this when we have `esModuleInterop` repo wide. I've had that on my radar to look at.

Comment (asklar) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:720]: this builds but fails at runtime during tests:
image

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (asklar) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:675]: i'll take a look in the refactor PR :)

File: packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts
Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:723]: Hmm... Maybe waiting until esModuleInterop it is enabled repo wide would be the best move after all.

If you want to use it right now, you should still be able to call the function from the require. Might need to console.log the shape of the returned object to find the property you need.

Comment (NickGerleman) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts:723]: Oh wait, turns out you can just require it without anything special.

image

Comment (asklar) [reply packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts]: thanks that seems to work

File: packages/@react-native-windows/cli/src/e2etest/projects/WithWinUI3/react-native.config.js
Comment (NickGerleman) [inline packages/@react-native-windows/cli/src/e2etest/projects/WithWinUI3/react-native.config.js:55]: Heads up that https://github.com/microsoft/react-native-windows/pull/7075 moves these to [jest snapshots](https://jestjs.io/docs/en/snapshot-testing). This lets you automatically update masters by running `yarn test -u` instead of needing to try to copy/paste from failing test output or manually editing.

**Pull request merged**
